### PR TITLE
Mutation tests

### PR DESCRIPTION
**Pull request by snapshotpl**

### CODE REVIEW

**Comment by xtreamwayz** [comment]
It makes more sense to do this for the develop branch since a lot is changed in that branch.

File: test/AppFactoryTest.php [hunk 1]
```diff
@@ -9,6 +9,20 @@
 
 namespace ZendTest\Expressive;
 
+namespace Zend\Expressive;
+
+use ZendTest\Expressive\AppFactoryTest;
+
+function class_exists($classname)
```
Comment (weierophinney) [inline test/AppFactoryTest.php]: This should be extracted into a separate file; I can do that on merge.

File: src/Application.php [hunk 2]
```diff
@@ -188,7 +196,7 @@ public function __call($method, $args)
         }
 
         // @TODO: we can use variadic parameters when dependency is raised to PHP 5.6
```
Comment (weierophinney) [inline src/Application.php]: Note to self: update the code to use variadics when merging.

File: test/ApplicationTest.php [hunk 3]
```diff
@@ -268,6 +268,17 @@ public function testThrowsExceptionWhenDispatchingUncallableMiddleware($middlewa
         });
     }
 
+    public function testExecuteMiddlewareIfNotRouteMatch()
```
Comment (weierophinney) [inline test/ApplicationTest.php]: Note to self: rename this to `testDispatchMiddlewareExecutesNextMiddlewareInStackIfNoRouteMatchPresent`.

File: test/Emitter/EmitterStackTest.php [hunk 4]
```diff
@@ -110,16 +110,31 @@ public function testEmitLoopsThroughEmittersUntilOneReturnsNonFalseValue()
         $second->emit(Argument::type(ResponseInterface::class))
             ->willReturn(null);
 
-        $third = $this->prophesize(EmitterInterface::class);
-        $third->emit(Argument::type(ResponseInterface::class))
-            ->willReturn(false);
```
Comment (weierophinney) [inline test/Emitter/EmitterStackTest.php]: The above was in the test for a specific reason: to ensure that the stack does not execute the third emitter once the second returns a non-false value. Perhaps the better solution would be to have it raise an exception:

```php
->will(function () {
    $this->fail('Should not reach third emitter');
})
```

I can make that change on merge.

File: src/Application.php (hunk 2 above)
Comment (weierophinney) [inline src/Application.php]: Interestingly, already done there, which means this was out-of-date.

File: test/Emitter/EmitterStackTest.php (hunk 4 above)
Comment (weierophinney) [inline test/Emitter/EmitterStackTest.php]: Forgot this was a stack. The actual workflow is:

- Hits `$third`, which returns `false`.
- Hits `$second`, which returns `null`, and should halt execution of the stack.
- Should **not** hit `$first`.

So, the point is to note that the first non-`false` value will stop execution of the stack. We want to ensure that we can go more than one deep in the stack, and that it does not just halt on the first encountered.

**Comment by weierophinney** [comment]
Thanks, @snapshotpl\!

**Pull request merged**